src/reducers/log.ts

```diff
@@ -0,0 +1,42 @@
+import {Action} from '../actions/index';
+import {LOG_ERRORS_ADD, LOG_ERRORS_REMOVE, LOG_WARNINGS_ADD, LOG_WARNINGS_REMOVE} from '../actions/log';
+import {isSpecAction} from '../actions/shelf/spec';
+import {DEFAULT_LOG, Log} from '../models/log';
+import {addNoDuplicateArrays} from './util';
+
+
+export function logReducer(log: Log = DEFAULT_LOG, action: Action): Log {
+  if (isSpecAction(action)) {
+    return {
+      errors: [],
+      warnings: []
+    };
+  }
+
+  switch (action.type) {
+    case LOG_ERRORS_ADD:
+      const {errors} = action.payload;
+      return {
+        ...log,
+        errors: addNoDuplicateArrays(log.errors, errors)
+      };
+    case LOG_WARNINGS_ADD:
+      const {warnings} = action.payload;
+      return {
+        ...log,
+        warnings: addNoDuplicateArrays(log.warnings, warnings)
```

Comment:
This will always re-create log object no matter we have new warning added or not.